backend/Application/Database/MigrationJobs/_03_UpdateFinalizationTimes.cs

```diff
@@ -0,0 +1,161 @@
+            .WrapAsync(rpcPolicy)
+            .ExecuteAsync(async () =>
+            {
+                var state = await _importStateController.GetStateIfExists();
+                if (state == null)
+                {
+                    return;
+                }
+
+                var initialValues = await FindInitialState(token);
+        
+                var (nextHeight, priorFinalizedBlockHeight, priorFinalizedBlockHash, finalHeight) = initialValues;
+                var initialHeight = nextHeight;
+                while (nextHeight < finalHeight)
+                {
+                    nextHeight++;
+                    _logger.Debug($"{JobName} processing block height {nextHeight}");
+                    var nextBlockInfo = await _client.GetBlockInfoAsync(new Absolute((ulong)nextHeight), token);
+
+                    var nextBlockLastFinalizedHash = nextBlockInfo.BlockLastFinalized.ToString();
+
+                    // Last finalization block hasn't changed
+                    if (nextBlockLastFinalizedHash == priorFinalizedBlockHash)
+                    {
+                        continue;
+                    }
+
+                    await using var context = await _contextFactory.CreateDbContextAsync(token);
+                    var finalizedBlock = await context.Blocks
+                        .SingleAsync(b => b.BlockHash == nextBlockLastFinalizedHash, token);
+
+                    var timeUpdate = new FinalizationTimeUpdate(priorFinalizedBlockHeight, finalizedBlock.BlockHeight);
+                    _logger.Debug($"Updating block height from:{timeUpdate.MinBlockHeight}, to {timeUpdate.MaxBlockHeight}");
+                    
+                    await BlockWriter.UpdateFinalizationTimes(timeUpdate, nextBlockInfo.BlockSlotTime, context);
+                    await _metricsWriter.UpdateFinalizationTimes(timeUpdate);
+
+                    priorFinalizedBlockHash = finalizedBlock.BlockHash;
+                    priorFinalizedBlockHeight = finalizedBlock.BlockHeight;
+                    ApplicationMetrics.SetJobCompletion(this, (nextHeight - initialHeight) / (double)finalHeight);
+                }
+            });
+        _logger.Debug($"Done processing {JobName}");
+    }
+
+    private sealed record InitialState(
+        int NextHeight,
+        int PriorFinalizedBlockHeight,
+        string PriorFinalizedBlockHash,
+        int FinalHeight
+    );
+
+    private async Task<InitialState> FindInitialState(CancellationToken token)
+    {
+        await using var context = await _contextFactory.CreateDbContextAsync(token);
+
+        var firstBlockWithNull = await context
+            .Blocks
+            .Where(b => b.BlockStatistics.FinalizationTime == null)
+            .OrderBy(b => b.BlockHeight)
+            .Take(1)
+            .SingleAsync(cancellationToken: token);
+            
+        var priorBlockInfo = await _client.GetBlockInfoAsync(new Absolute((ulong)firstBlockWithNull.BlockHeight - 1), token);
```

Comment:
Do we need to worry about the block height being 0?

Reply:
no since we know the data in the databases.